Improve error messages when `@apply` fails

RobinMalfait opened the pull request:
This PR improves error messages when `@apply` fails. Right now it gives you a generic error message that you cannot apply a certain utility.

```css
.foo {
  @apply bg-red-500;
}
```

Would result in:
```
Cannot apply unknown utility class: bg-red-500
```

However, there are some situations where we can give you more context about what's happening.

### Missing `@import "tailwindcss"` or `@reference`

If you are in a Vue file for example, and you have the following code:
```vue
<template>
  <div class="foo"></div>
</template>

<style>
.foo {
  @apply bg-red-500;
}
</style>
```

Then this will now result in:
```
Cannot apply unknown utility class `bg-white`. Are you using CSS modules or similar and missing `@reference`? https://tailwindcss.com/docs/functions-and-directives#reference-directive
```

We do this by checking if we found a `@tailwind utilities` or `@reference`. If not, we throw this more specific error.

### Explicitly excluded classes via `@source not inline('…')`

Or via the legacy `blocklist` from a JS config.

If you then have the following file:
```css
@import "tailwindcss";

@source not inline('bg-white');

.foo {
  @apply bg-white;
}
```

Then this will now result in:
```
Cannot apply utility class `bg-white` because it has been explicitly disabled: https://tailwindcss.com/docs/detecting-classes-in-source-files#explicitly-excluding-classes
```

We do this by checking if the class was marked as invalid.

### Applying unprefixed class in prefix mode

If you have the prefix option configured, but you are applying a non-prefixed class, then we will show the following error:

Given this input:
```css
@import "tailwindcss" prefix(tw);

.foo {
  @apply underline;
}
```


The following error is thrown:
```
Cannot apply unprefixed utility class `underline`. Did you mean `tw:underline`?
```

### Applying known utilities with unknown variants

If you have unknown variants, then we will list them as well if the base utility does compile correctly.

Given this input:

```css
@import "tailwindcss";

.foo {
  @apply hocus:hover:pocus:bg-red-500;
}
```

The following error is thrown:
```
Cannot apply utility class `hocus:hover:pocus:bg-red-500` because the `hocus` and `pocus` variants do not exist.
```

## Test plan

1. Everything behaves the same, but the error messages give more details.
2. Updated tests with new error messages
3. Added new unit tests to verify the various scenarios
4. Added a Vue specific integration test with a `<style>…</style>` block using `@apply`

[ci-all] There are some newlines here and there, let's verify that they work identically on all platforms.

RobinMalfait committed:
improve `@apply` error messages

RobinMalfait committed:
add and update tests

RobinMalfait committed:
delay making `@tailwind utilities` a context node

If we do this _before_ handling `@apply`, then we won't be able to find
a `@tailwind utilities` at-rule.

RobinMalfait committed:
add `@apply` integration tests for Vue `<style>` blocks

RobinMalfait committed:
normalize whitespace

RobinMalfait committed:
are there VT control characters? Who knows!

thecrypticace commented on packages/tailwindcss/src/apply.ts:
```diff
@@ -176,7 +176,54 @@ export function substituteAtApply(ast: AstNode[], designSystem: DesignSystem) {
+              throw new Error(
+                `Cannot apply \`${candidate}\`, it seems like the utility was explicitly excluded and cannot be applied.\n\nMore info: https://tailwindcss.com/docs/detecting-classes-in-source-files#explicitly-excluding-classes`,
+              )
+            }
```
This is a nice feature 💯

thecrypticace reviewed (approved):
Left a note about the stylesheet name but I'm gonna go ahead and approve this. We can tweak those details later if need be.

RobinMalfait commented on packages/tailwindcss/src/apply.ts:
```diff
@@ -176,7 +176,54 @@ export function substituteAtApply(ast: AstNode[], designSystem: DesignSystem) {
+            })
+            if (!hasUtilitiesOrReference) {
+              throw new Error(
+                `Cannot apply unknown utility class: \`${candidate}\`.\nIt looks like you are missing a \`@reference "app.css"\` or \`@import "tailwindcss";\``,
```
I used `app.css` because we use that a lot in the docs, e.g.: https://tailwindcss.com/docs/theme#what-are-theme-variables

thecrypticace replied:
Ah that's fair. Seems fine them.

RobinMalfait committed:
update changelog

philipp-spiess commented on packages/tailwindcss/src/apply.ts:
```diff
@@ -176,7 +176,54 @@ export function substituteAtApply(ast: AstNode[], designSystem: DesignSystem) {
+              else if (
+                node.kind === 'at-rule' &&
+                node.name === '@tailwind' &&
+                (node.params === 'utilities' || node.params.startsWith('utilities'))
```
Let's check the `designSystem` itself so it:

- works when there's only a theme file inlined
- we can check it in constant time
- it will also work if the theme config is only coming from legacy plugins

RobinMalfait committed:
rely on `Theme` to know when we are missing `@reference` or `@import "tailwindcss"`

This is a better single source of truth because `Theme` will be filled
in with the necessary data.

RobinMalfait committed:
show unknown variants when using `@apply`

reinink commented on packages/tailwindcss/src/apply.ts:
```diff
@@ -176,7 +177,60 @@ export function substituteAtApply(ast: AstNode[], designSystem: DesignSystem) {
+            // very likely missing.
+            if (designSystem.theme.size === 0) {
+              throw new Error(
+                `Cannot apply unknown utility class: \`${candidate}\`.\nIt looks like you are missing a \`@reference "app.css"\` or \`@import "tailwindcss";\``,
```
```suggestion
                `Cannot apply \`${candidate}\`, as this class is unknown. It looks like you are missing \`@import "tailwindcss";\` or \`@reference "app.css"\``,
```

```txt
Cannot apply `bg-white`, as this class is unknown. It looks like you are missing either `@import "tailwindcss";` or `@reference "app.css";` in your project.
```

reinink commented on packages/tailwindcss/src/apply.ts:
```diff
@@ -176,7 +177,60 @@ export function substituteAtApply(ast: AstNode[], designSystem: DesignSystem) {
+            }
+
+            // Fallback to most generic error message
+            throw new Error(`Cannot apply unknown utility class: \`${candidate}\``)
```
```suggestion
            throw new Error(`Cannot apply \`${candidate}\`, as this class is unknown.`)
```

```txt
Cannot apply `bg-white`, as this class is unknown.
```

reinink reviewed (dismissed):
Some feedback — still needs to be reviewed by @adamwathan, so don't merge anything in yet.

RobinMalfait committed:
Apply suggestions from code review

Co-authored-by: Jonathan Reinink <jonathan@reinink.ca>

RobinMalfait committed:
adjust error messages
